### PR TITLE
refactor: plugin.diff.file|type -> plugin.detectChanges

### PR DESCRIPTION
**Pull request by samuelstroschein**
- simpler API
- no redundant reporting of before and after

closes https://github.com/opral/lix-sdk/issues/40

### CODE REVIEW

File: lix/packages/sdk/src/file-handlers.ts [hunk 1]
```diff
@@ -19,54 +20,58 @@ export async function handleFileInsert(args: {
...
-					})
-					.execute();
-			}
+		for (const detectedChange of detectedChanges) {
+			const snapshot = await trx
+				.insertInto("snapshot")
+				.values({
+					// TODO use empty snapshot id const https://github.com/opral/lix-sdk/issues/101
+					id: detectedChange.snapshot ? undefined : "EMPTY_SNAPSHOT_ID",
+					// @ts-expect-error - database expects stringified json
```
Comment (samuelstroschein) [inline lix/packages/sdk/src/file-handlers.ts:57]: Assuming here that EMPTY_SNAPSHOT_ID will come with https://github.com/opral/lix-sdk/issues/101

File: lix/packages/sdk/src/resolve-conflict/resolve-conflict-with-new-change.test.ts [hunk 2]
```diff
@@ -137,15 +133,11 @@ test("resolving a conflict should throw if the to be resolved with change is not
 
 	const mockPlugin: LixPlugin = {
 		key: "plugin1",
-		glob: "*",
```
Comment (samuelstroschein) [inline lix/packages/sdk/src/resolve-conflict/resolve-conflict-with-new-change.test.ts:140]: Removing all redundant properties that are not tested.

**Comment by samuelstroschein** [comment]
@martin-lysk straightforward change after our sync. no explicit review request. if you find something lmk

**Comment by samuelstroschein** [comment]
@martin-lysk if you reviewed and merges this, go ahead and merge #3177 too which builds on this PR

**Review by martin-lysk — COMMENTED** [review]
I got to review some changes in the tests but here is the WIP review

File: lix/packages/sdk/src/file-handlers.ts (hunk 1 above)
Comment (samuelstroschein) [reply lix/packages/sdk/src/file-handlers.ts:57]: will be handled in follow up snapshot content addressable pr

File: lix/packages/sdk/src/file-handlers.ts [hunk 3]
```diff
@@ -87,97 +92,99 @@ export async function handleFileChange(args: {
...
+			const previousChanges = await trx
+				.selectFrom("change")
+				.selectAll()
+				.where("file_id", "=", fileId)
+				.where("plugin_key", "=", detectedChange.pluginKey)
+				.where("type", "=", detectedChange.type)
+				.where("entity_id", "=", detectedChange.entity_id)
+				.execute();
+
+			// we need to finde the real leaf change as multiple changes can be set in the same created timestamp second or clockskew
```
[the remark this answers is not in the document]
Comment (samuelstroschein) [reply lix/packages/sdk/src/file-handlers.ts:134]: yes

File: lix/packages/sdk/src/file-handlers.ts [hunk 4]
```diff
@@ -87,97 +92,99 @@ export async function handleFileChange(args: {
...
+			}
 
-				const snapshot = await trx
-					.insertInto("snapshot")
-					.values({
-						// @ts-expect-error - database expects stringified json
-						value: JSON.stringify(value),
-					})
-					.returning("id")
+			// working change exists but is identical to previously committed change
```
[the remark this answers is not in the document]
Comment (samuelstroschein) [reply lix/packages/sdk/src/file-handlers.ts]: solved

**Pull request merged**